### PR TITLE
dpctgan synthesizer

### PR DESCRIPTION
**Pull request by AprilXiaoyanLiu**
- add Differential Privacy implementation to [CTGAN](https://github.com/sdv-dev/CTGAN) 
- modify torchdp to support multi loss.backward() 
- TO-DO: modify _custom_create_or_extend_grad_sample function to consider scenario of using optimizer.virtual_step()

### CODE REVIEW

File: sdk/opendp/whitenoise/synthesizers/dpctgan/dpctgan.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+) -> None:
+    """
+    Create a 'grad_sample' attribute in the given parameter, or accumulate it
+    if the 'grad_sample' attribute already exists.
+
+    This custom code will not work when using optimizer.virtual_step()
+    """
+    
+    if hasattr(param, "grad_sample"):
+        param.grad_sample = param.grad_sample + grad_sample
```
Comment (eedeleon) [inline sdk/opendp/whitenoise/synthesizers/dpctgan/dpctgan.py:34]: can we look into the approach mentioned where we add the virtual_step attribute for the override to avoid affecting base scenarios?

if hasattr(param, "grad_sample") and getattr(param, "virtual_step", False):
           param.grad_sample = param.grad_sample + grad_sample
  elif if hasattr(param, "grad_sample"):
       param.grad_sample = torch.cat((param.grad_sample, grad_sample), batch_dim)
  else:
        param.grad_sample = grad_sample

File: sdk/opendp/whitenoise/synthesizers/dpctgan/dpctgan.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+import ctgan
+from ctgan.transformer import DataTransformer
+from ctgan.conditional import ConditionalGenerator 
+from ctgan.models import Discriminator, Generator
+from ctgan.sampler import Sampler 
+
+from ctgan import CTGANSynthesizer
+
+import torchdp
```
Comment (eedeleon) [inline sdk/opendp/whitenoise/synthesizers/dpctgan/dpctgan.py:19]: We should move this into a torchdp directory, and have that directory pip install torchdp

 sdk/opendp/whitenoise/synthesizers/torchdp/

@anujdesai510  since you will probably need to do something similar, this can be how we deal with the optional dependency for now

File: tests/sdk/synthesizers/test_dpctgan.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+meta_path = os.path.join(git_root_dir, os.path.join("service", "datasets", "PUMS.yaml"))
+csv_path = os.path.join(git_root_dir, os.path.join("service", "datasets", "PUMS.csv"))
+
+schema = CollectionMetadata.from_file(meta_path)
+df = pd.read_csv(csv_path)
+
+synth = DPCTGANSynthesizer()
+
+class TestDPCTGAN:
+    def test_fit(self):
```
Comment (eedeleon) [inline tests/sdk/synthesizers/test_dpctgan.py:22]: nice!

File: sdk/opendp/whitenoise/synthesizers/dpctgan/dpctgan.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+from torch.nn import functional as F
+from opendp.whitenoise.synthesizers.base import SDGYMBaseSynthesizer
+
+import ctgan
+from ctgan.transformer import DataTransformer
+from ctgan.conditional import ConditionalGenerator 
+from ctgan.models import Discriminator, Generator
+from ctgan.sampler import Sampler 
+
+from ctgan import CTGANSynthesizer
```
Comment (eedeleon) [inline sdk/opendp/whitenoise/synthesizers/dpctgan/dpctgan.py:17]: what library is ctgan?
